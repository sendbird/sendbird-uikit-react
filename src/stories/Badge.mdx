import { Meta, Story, Canvas, Controls, ArgsTable } from '@storybook/addon-docs';
import Badge from '../ui/Badge';
import * as BadgeStories from './Badge.stories';

<Meta title="UI/Badge" component={Badge} of={BadgeStories} />

# Badge

The `Badge` component is designed to display numerical values or indicators, such as unread message counts. It can be customized to show different levels of counts and can adapt to various UI themes with custom class names.
Utilize the Badge component to effectively represent numerical values or indicators within your user interface, enhancing the informational clarity of your application.

<Canvas of={BadgeStories.Default} />
<Controls />

### Importing Badge

To use the `Badge` component in your project, import it as follows:

```tsx
import Badge from '@sendbird/uikit-react/ui/Badge';
```

### Usage
Here's how to use the Badge component in your application:

<Canvas of={BadgeStories.Default} />


With String Count
Demonstrates the Badge component with a string count, typically used for displaying counts beyond a certain number, like '99+'.

<Canvas of={BadgeStories.StringCount} />

With Maximum Level
Shows how to use the maxLevel prop to limit the maximum number of digits before showing an overflow indicator.

<Canvas of={BadgeStories.MaxLevelThree} />
