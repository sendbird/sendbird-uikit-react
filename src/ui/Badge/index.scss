@import '../../styles/variables';

.sendbird-badge {
  height: 20px;
  min-width: 20px;
  border-radius: 10px;
  display: inline-flex;
  justify-content: center;
  align-items: center;
  @include themed() {
    background-color: t(primary-3);
  }

  .sendbird-badge__text {
    margin: 0px 6px;
  }
}
