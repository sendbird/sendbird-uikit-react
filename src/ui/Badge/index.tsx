import React, { ReactElement, useContext } from 'react';

import './index.scss';
import Label, { LabelTypography, LabelColors } from '../Label';
import { LocalizationContext } from '../../lib/LocalizationContext';

export interface BadgeProps {
  count: number;
  maxLevel?: number;
  className?: string | Array<string>;
}

export default function Badge({
  count,
  maxLevel = 2,
  className = '',
}: BadgeProps): ReactElement {
  const { stringSet } = useContext(LocalizationContext);
  const maximumNumber = parseInt('9'.repeat((maxLevel > 6) ? 6 : maxLevel), 10);
  return (
    <div
      className={[
        ...(Array.isArray(className) ? className : [className]),
        'sendbird-badge',
      ].join(' ')}
    >
      <div className="sendbird-badge__text">
        <Label
          type={LabelTypography.CAPTION_2}
          color={LabelColors.ONCONTENT_1}
        >
          {
            (count > maximumNumber)
              ? `${maximumNumber}${stringSet.BADGE__OVER}`
              : count
          }
        </Label>
      </div>
    </div>
  );
}
