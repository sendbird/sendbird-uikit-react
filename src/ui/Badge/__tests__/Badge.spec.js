import React from 'react';
import { render, screen } from '@testing-library/react';

import Badge from "../index";

describe('Badge', () => {
  it('should do a snapshot test of the Badge DOM', function () {
    const count = 1;
    const { asFragment } = render(<Badge count={count} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should display maximum number as a maxLevel: 1', function () {
    const count = 10;
    const level = 1;

    render(<Badge count={count} maxLevel={level} />);
    expect(screen.getByText('9+')).toBeTruthy();
  });

  it('should display maximum number as a maxLevel: 2 (default level)', function () {
    const count = 100;

    render(<Badge count={count} />);
    expect(screen.getByText('99+')).toBeTruthy();
  });

  it('should display maximum number as a maxLevel: 3', function () {
    const count = 1000;
    const level = 3;

    render(<Badge count={count} maxLevel={level} />);
    expect(screen.getByText('999+')).toBeTruthy();
  });
});
