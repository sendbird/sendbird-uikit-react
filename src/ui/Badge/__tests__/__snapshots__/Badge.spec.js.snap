// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Badge should do a snapshot test of the Badge DOM 1`] = `
<div
  className=" sendbird-badge"
>
  <div
    className="sendbird-badge__text"
  >
    <span
      className="sendbird-label sendbird-label--caption-2 sendbird-label--color-oncontent-1"
    >
      1
    </span>
  </div>
</div>
`;
